### ISSUE TITLE
Return predicted trajectory coordinates to the map

### ISSUE THREAD

**Issue by mpourkeshavarz** [MediaBrain-SJTU/MemoNet#5]
Hi, @WBMao @sjtuxcx 
There is a function `visualize_data()` in map.py to get past and future trajectory coordinates back onto the map. Could you also explain how to return predicted trajectory coordinates to the map?
PS: this conversion would be equivalent to getting back **past_normalized** to the **past**.

```
	past = torch.stack(data['pre_motion_3D']).cuda()
	future = torch.stack(data['fut_motion_3D']).cuda()
	last_frame = past[:, -1:]
	past_normalized = past - last_frame
	fut_normalized = future - last_frame
	

	past_abs = past.unsqueeze(0).repeat(past.size(0), 1, 1, 1)
	past_centroid = past[:, -1:, :].unsqueeze(1)
	past_abs = past_abs - past_centroid

	scale = 1
	if self.cfg.scale.use:
		scale = torch.mean(torch.norm(past_normalized[:, 0], dim=1)) / 3
		if scale<self.cfg.scale.threshold:
			scale = 1
		else:
			if self.cfg.scale.type == 'divide':
				scale = scale / self.cfg.scale.large
			elif self.cfg.scale.type == 'minus':
				scale = scale - self.cfg.scale.large
		if self.cfg.scale.type=='constant':
			scale = self.cfg.scale.value
		past_normalized = past_normalized / scale
		past_abs = past_abs / scale

	if self.cfg.rotation:
		past_normalized, fut_normalized, past_abs = self.rotate_traj(past_normalized, fut_normalized, past_abs)
```

**Comment by sjtuxcx** [comment]
Hi, here we are not using the map information but pure trajectories. If you want to visualize the trajectory into maps, you may need to get the original video and make the coordinates aligned with the maps. Here we only use some translation and scaling, you can do the inverse operation to return them to the original data space.

**Comment by mpourkeshavarz** [comment]
Regardless of the map, I can visualize the past and the future trajectory(GT) but have some problems in scaling up the predicted trajectory to be comparable with the future trajectory(GT).

Would you explain how it could be possible to get back past_normalized to past (variables in the above code)? It's passed through 3 steps:
(past ~ pre_motion_3D) --> apply normalization using last_frame --> scale --> rotation
I guess if I can do the reverse step, I can do the same for the predicted trajectory.

**Comment by sjtuxcx** [comment]
Yes, if you want to get back the past from past_normalized, just do the following operation：inverse rotation --> rescale --> add the last frame. For the future trajectory, the operation is the same.

**Comment by sjtuxcx** [comment]
By the way, actually, the map.py file is not used. Maybe it brings you some confusion and we are sorry for that. We will delete this file soon.

**Comment by mpourkeshavarz** [comment]
Thanks for your answer @sjtuxcx ,
Maybe I should refine my question,
In any case, could you share the code for generating trajectories similar to those provided in figures 7, 8, and 9 of the paper, regardless of the map?

**Comment by SungwookLE** [comment]
@sjtuxcx Thank you for your working, Could you share the code for generating trajectories similar to figures that you made? could you share the visualization code?
